### ISSUE TITLE
Change allowArrowFunctions in func-style to string option

### ISSUE THREAD

**Issue by jhnns** [eslint/eslint#7765]
Hi everyone! Thanks for eslint, it is awesome :)

I would like to allow arrow functions without bodies, like:

```javascript
// Good
const bla = () => true;
// Good
const bla = () => ({
    prop: true
});
```

```javascript
// Bad
const bla = () => {
    return true;
};
// Bad
const bla = () => {
    const blub = true;
    return  blub;
};
```

You'd configure it like this:

```javascript
/*eslint func-style: ["error", "declaration", { "allowArrowFunctions": "expression-body" }]*/
/*eslint func-style: ["error", "declaration", { "allowArrowFunctions": "all" }]*/
```

What you be open for that change or do you see any problems?

@flootr would be interested in doing a PR :)

**Comment by eslintbot** [comment]
Hi @jhnns, thanks for the issue. It looks like there's not enough information for us to know how to help you.

If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

Requesting a rule change? Please see [Proposing a Rule Change](http://eslint.org/docs/developer-guide/contributing/rule-changes) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by jhnns** [comment]
**What rule do you want to change?**
func-style

**Does this change cause the rule to produce more or fewer warnings?**
The default option is not changed.

**How will the change be implemented? (New option, new default behavior, etc.)?**
Instead of `allowArrowFunction` being a boolean option, you'd use a string to specify what kind of arrow function would be allowed:

- `never` (default)
- `without-body`
- `always`

**Please provide some example code that this change will affect:**

```js
/*eslint func-style: ["error", "declaration", { "allowArrowFunctions": "never" }]*/
// Good
function bla() {}
// Bad
var bla = function () {}
var bla = () => {}
var bla = () => true
```

```js
/*eslint func-style: ["error", "declaration", { "allowArrowFunctions": "without-body" }]*/
// Good
function bla() {}
var bla = () => true
// Bad
var bla = function () {}
var bla = () => {}
```

```js
/*eslint func-style: ["error", "declaration", { "allowArrowFunctions": "always" }]*/
// Good
function bla() {}
var bla = () => true
var bla = () => {}
// Bad
var bla = function () {}
```

```js
/*eslint func-style: ["error", "expression", { "allowArrowFunctions": "never" }]*/
// Good
var bla = function () {}
// Bad
function bla() {}
var bla = () => true
var bla = () => {}
```

```js
/*eslint func-style: ["error", "expression", { "allowArrowFunctions": "without-body" }]*/
// Good
var bla = function () {}
var bla = () => true
// Bad
function bla() {}
var bla = () => {}
```

```js
/*eslint func-style: ["error", "expression", { "allowArrowFunctions": "always" }]*/
// Good
var bla = function () {}
var bla = () => true
var bla = () => {}
// Bad
function bla() {}
```

**What does the rule currently do for this code?**
Currently I can only allow or disallow arrow functions.

**What will the rule do after it's changed?**
It will provide a more fine-grained control over what kind of arrow functions are allowed.

In our case we wanted to enforce the declaration style while still allowing this special kind of arrow functions since they are so common in functional programming (and much shorter than the declaration style).

**Comment by nzakas** [comment]
@eslint/eslint-team looks like we lost track of this. Any thoughts?

**Comment by btmills** [comment]
@jhnns is your goal to permit expression-bodied arrow functions everywhere, or just in certain places (e.g. async callbacks and array iterators like `forEach` and `map`)?

**Comment by jhnns** [comment]
Our goal was this:

We generally use and enforce function declarations. However, if you're doing a lot of functional programming, it is very common to have functions that immediately return something. In this case, writing

```js
function newUser() {
    return {
        name: "john"
    };
}
```

feels more cumbersome than

```js
const newUser = () => ({
    name: "john"
});
```

I felt like that this might be a reasonable exception to deviate from the declaration rule.

**Comment by btmills** [comment]
I've added my :+1:. I would, however, like to suggest `"never"`/`"expression-body"`/`"always"` as the option names, instead of `"without-body"` for the in-between option.

**Comment by jhnns** [comment]
Yeah, `expression-body` is better. I've updated my comments.

@flootr do you want to tackle this? @btmills can help you if you're stuck.

**Comment by platinumazure** [comment]
I'm :+1: for this. Now we just need a champion from the team to move this forward.

**Comment by ilyavolodin** [comment]
This proposal now has enough votes. Anyone from the core team wants to champion this?

**Comment by not-an-aardvark** [comment]
@eslint/eslint-team This proposal has a lot of votes. Is anyone willing to champion it?

**Comment by kaicataldo** [comment]
Maybe I'm missing something, but why does this change need to happen in `func-style` instead of `arrow-body-style`? Seems to me like we're mixing responsibilities here.

**Comment by btmills** [comment]
@jhnns building on what @kaicataldo suggested, could you accomplish this today with these two existing rules?

```json
{
    "rules": {
        "arrow-body-style": ["error", "never"],
        "func-style": ["error", "declaration", {
            "allowArrowFunctions": true
        }]
    }
}
```

**Comment by not-an-aardvark** [comment]
Closing this issue because no one on the team was willing to champion it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after a long time tend to never do it, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
